Bug 1237407 - Add ability to specify custom FxA servers

vbudhram opened the pull request:
This PR adds the ability to specify custom FxA servers and sync via debug option. To access click `Advance Account Settings` while in debug mode. To test, I used `https://latest.dev.lcip.org`. 

Some quick notes
 * Uses FxAs `/.well-known/fxa-client-configuration/` to pull in those servers configuration
 * Signs out the account whenever you change servers
 * Stores custom server configuration in user prefs

![screen shot 2017-08-23 at 1 32 08 pm](https://user-images.githubusercontent.com/1295288/29629618-80055f20-8807-11e7-811c-c6428c36df92.png)


Fixes https://github.com/mozilla/fxa-bugzilla-mirror/issues/332

farhanpatel commented:
<!--
  0 Errors
  0 Warnings
  0 Messages
  1 Markdown
-->

### SwiftLint found issues

#### Warnings

File | Line | Reason |
| --- | ----- | ----- |
AppSettingsOptions.swift | 715 | Limit vertical whitespace to a single empty line. Currently 2. 
AppSettingsOptions.swift | 258 | Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. 
AppSettingsOptions.swift | 187 | Arguments can be omitted when matching enums with associated types if they are not used. 
Profile.swift | 757 | Prefer `() -> ` over `Void -> `. 
#### Errors

File | Line | Reason |
| --- | ----- | ----- |
Profile.swift | 478 | Opening braces should be preceded by a single space and on the same line as the declaration. 


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

vbudhram commented:
@st3fan Is it suffice for the feature flag for this to be debug mode option?

st3fan commented:
@vbudhram No, enable this for all builds please. I think it will be very useful if Release and Beta can also use this.

jhugman commented on Account/FirefoxAccountConfiguration.swift:
```diff
@@ -27,7 +30,7 @@ public enum FirefoxAccountConfigurationLabel: String {
  */
 public protocol FirefoxAccountConfiguration {
-    init()
+    init(prefs: Prefs?)
```
Remove both of these lines, then you can just use the constructor you need for each type (you don't have to use the same one for all types).

jhugman commented on Account/FirefoxAccountConfiguration.swift:
```diff
@@ -53,7 +56,7 @@ public protocol FirefoxAccountConfiguration {
 
 public struct LatestDevFirefoxAccountConfiguration: FirefoxAccountConfiguration {
-    public init() {
+    public init(prefs: Prefs? = nil) {
```
This change isn't needed if you remove the constructor from the `FirefoxAccountConfiguration`.

jhugman commented on Account/FirefoxAccountConfiguration.swift:
```diff
@@ -72,7 +75,7 @@ public struct LatestDevFirefoxAccountConfiguration: FirefoxAccountConfiguration
 
 public struct StableDevFirefoxAccountConfiguration: FirefoxAccountConfiguration {
-    public init() {
+    public init(prefs: Prefs? = nil) {
```
This change isn't needed if you remove the constructor from the `FirefoxAccountConfiguration`.

jhugman commented on Account/FirefoxAccountConfiguration.swift:
```diff
@@ -91,7 +94,7 @@ public struct StableDevFirefoxAccountConfiguration: FirefoxAccountConfiguration
 
 public struct StageFirefoxAccountConfiguration: FirefoxAccountConfiguration {
-    public init() {
+    public init(prefs: Prefs? = nil) {
```
This change isn't needed if you remove the constructor from the `FirefoxAccountConfiguration`.

jhugman commented on Account/FirefoxAccountConfiguration.swift:
```diff
@@ -120,7 +123,7 @@ public struct StageFirefoxAccountConfiguration: FirefoxAccountConfiguration {
 
 public struct ProductionFirefoxAccountConfiguration: FirefoxAccountConfiguration {
-    public init() {
+    public init(prefs: Prefs? = nil) {
```
This change isn't needed if you remove the constructor from the `FirefoxAccountConfiguration`.

jhugman commented on Account/FirefoxAccountConfiguration.swift:
```diff
@@ -138,8 +141,83 @@ public struct ProductionFirefoxAccountConfiguration: FirefoxAccountConfiguration
+
 public struct ChinaEditionFirefoxAccountConfiguration: FirefoxAccountConfiguration {
-    public init() {
+    public init(prefs: Prefs? = nil) {
```
This change isn't needed if you remove the constructor from the `FirefoxAccountConfiguration`.

jhugman commented on Account/FirefoxAccountConfiguration.swift:
```diff
@@ -157,28 +235,46 @@ public struct ChinaEditionFirefoxAccountConfiguration: FirefoxAccountConfigurati
-    public init() {
-    }
+public protocol Sync15Configuration {
+    init(prefs: Prefs?)
```
Remove this constructor.

jhugman commented on Account/FirefoxAccountConfiguration.swift:
```diff
@@ -157,28 +235,46 @@ public struct ChinaEditionFirefoxAccountConfiguration: FirefoxAccountConfigurati
-    init()
-    var tokenServerEndpointURL: URL { get }
+public struct ChinaEditionSync15Configuration: Sync15Configuration {
+    public init(prefs: Prefs? = nil) {
```
This change isn't needed if you remove the constructor from the `Sync15Configuration`.

jhugman commented on Account/FirefoxAccountConfiguration.swift:
```diff
@@ -157,28 +235,46 @@ public struct ChinaEditionFirefoxAccountConfiguration: FirefoxAccountConfigurati
 
 public struct ProductionSync15Configuration: Sync15Configuration {
-    public init() {
+    public init(prefs: Prefs? = nil) {
```
This change isn't needed if you remove the constructor from the `Sync15Configuration`.

jhugman commented on Account/FirefoxAccountConfiguration.swift:
```diff
@@ -157,28 +235,46 @@ public struct ChinaEditionFirefoxAccountConfiguration: FirefoxAccountConfigurati
 
 public struct StageSync15Configuration: Sync15Configuration {
-    public init() {
+    public init(prefs: Prefs? = nil) {
```
This change isn't needed if you remove the constructor from the `Sync15Configuration`.

jhugman commented on Client.xcodeproj/project.pbxproj:
```diff
@@ -4391,7 +4394,7 @@
 					3BFE4B061D342FB800DDF53F = {
 						CreatedOnToolsVersion = 7.3.1;
-						DevelopmentTeam = 43AQ936H96;
+						DevelopmentTeam = 9G8J6YA743;
```
Please don't change the `DevelopmentTeam`.

vbudhram replied:
Strange, didn't realize that changed, will revert.

jhugman commented on Client.xcodeproj/project.pbxproj:
```diff
@@ -4405,7 +4408,7 @@
 					D39FA15E1A83E0EC00EE869C = {
 						CreatedOnToolsVersion = 6.1.1;
-						DevelopmentTeam = 43AQ936H96;
+						DevelopmentTeam = 9G8J6YA743;
```
Please don't change the `DevelopmentTeam`.

jhugman commented on Client/Frontend/Settings/AdvanceAccountSettingViewController.swift:
```diff
@@ -0,0 +1,158 @@
+    }
+    
+    override func generateSettings() -> [SettingSection] {
+        let prefs = profile.prefs
```
You should probably be using a `BoolSetting` to toggle this on and off.

Changing the prefs only on *Set* is a bit weird from a UX. Have a look at how Home Page settings work.

jhugman commented on Client/Frontend/Settings/AppSettingsTableViewController.swift:
```diff
@@ -126,9 +126,10 @@ if
                 ExportBrowserDataSetting(settings: self),
                 DeleteExportedDataSetting(settings: self),
                 EnableBookmarkMergingSetting(settings: self),
+                AdvanceAccountSetting(settings: self),
```
Can we move this further up: e.g. below `ConnectSetting`?

vbudhram commented:
@jhugman Thanks review, I updated per comments. Mind another r?

Updated screens:
<img width="298" alt="screen shot 2017-08-30 at 2 45 12 pm" src="https://user-images.githubusercontent.com/1295288/29889366-ed461266-8d91-11e7-9aa5-24aa313e00c7.png">

jhugman commented on Account/FirefoxAccountConfiguration.swift:
```diff
@@ -74,7 +75,7 @@ public struct LatestDevFirefoxAccountConfiguration: FirefoxAccountConfiguration
     public init() {
     }
-
+    
```
Nit: whitespace

jhugman commented on Account/FirefoxAccountConfiguration.swift:
```diff
@@ -93,7 +94,7 @@ public struct StableDevFirefoxAccountConfiguration: FirefoxAccountConfiguration
     public init() {
     }
-
+    
```
Nit: whitespace

jhugman commented on Account/FirefoxAccountConfiguration.swift:
```diff
@@ -122,7 +123,7 @@ public struct StageFirefoxAccountConfiguration: FirefoxAccountConfiguration {
     public init() {
     }
-
+    
```
Nit: whitespace

jhugman commented on Client/Frontend/Settings/AdvanceAccountSettingViewController.swift:
```diff
@@ -0,0 +1,177 @@
+                                                         prefKey: PrefsKeys.KeyCustomSyncWeb,
+                                                         placeholder: Strings.SettingsAdvanceAccountUrlPlaceholder,
+                                                         accessibilityIdentifier: "CustomSyncSetting",
+                                                         settingDidChange: {fieldText in
```
Nit: whitespace

```swift
{ fieldText in
```

jhugman commented on Client/Frontend/Settings/AdvanceAccountSettingViewController.swift:
```diff
@@ -0,0 +1,177 @@
+                                                         accessibilityIdentifier: "CustomSyncSetting",
+                                                         settingDidChange: {fieldText in
+                                                            self.customSyncUrl = fieldText
+                                                            if self.customSyncUrl == "" {
```
Nit: use `.isEmpty` instead of ` == ""`

jhugman commented on Client/Frontend/Settings/AppSettingsOptions.swift:
```diff
@@ -821,8 +820,14 @@ class StageSyncServiceDebugSetting: WithoutAccountSetting {
+        let isCustomSync = prefs.boolForKey(PrefsKeys.KeyUseCustomSyncService) ?? false
+        
+        var configurationURL = isCustomSync ? CustomFirefoxAccountConfiguration(prefs: profile.prefs).authEndpointURL : ProductionFirefoxAccountConfiguration().authEndpointURL
+        configurationURL = isOn ? StageFirefoxAccountConfiguration().authEndpointURL : configurationURL
```
Swiftyness: Rewrite this as an `if`.

jhugman commented on Client/Frontend/Settings/AppSettingsOptions.swift:
```diff
@@ -902,3 +907,36 @@ class OpenWithSetting: Setting {
+    }
+    
+    override var hidden: Bool {
+        if !ShowDebugSettings {
```
This is disgusting, but not your fault. No action required in this patch.

jhugman commented on Client/Frontend/Settings/AppSettingsOptions.swift:
```diff
@@ -902,3 +907,36 @@ class OpenWithSetting: Setting {
+            return true
+        }
+        return false
+    }
```
```swift
override var hidden: Bool {
    return !ShowDebugSettings || profile.hasAccount()
}
```

jhugman commented on Client/Frontend/Settings/AppSettingsTableViewController.swift:
```diff
@@ -46,7 +46,7 @@ class AppSettingsTableViewController: SettingsTableViewController {
             BoolSetting(prefs: prefs, prefKey: "saveLogins", defaultValue: true,
                         titleText: NSLocalizedString("Save Logins", comment: "Setting to enable the built-in password manager")),
-            ]        
+            ]
```
Nit: whitespace

jhugman commented on Client/Frontend/Settings/AppSettingsTableViewController.swift:
```diff
@@ -128,7 +129,7 @@ if
                 ForceCrashSetting(settings: self)
             ])]
-    
+
```
Nit: whitespace

jhugman commented on Client/Frontend/Settings/AppSettingsTableViewController.swift:
```diff
@@ -33,7 +33,7 @@ class AppSettingsTableViewController: SettingsTableViewController {
             RequireUpgradeDebugSetting(settings: self),
             ForgetSyncAuthStateDebugSetting(settings: self),
-            StageSyncServiceDebugSetting(settings: self),
+            StageSyncServiceDebugSetting(settings: self)
```
Nit: Keep the trailing comma.

jhugman commented on Providers/Profile.swift:
```diff
@@ -473,20 +473,32 @@ open class BrowserProfile: Profile {
-        if prefs.boolForKey("useStageSyncService") ?? false {
-            return StageFirefoxAccountConfiguration()
+        let useCustomSyncService = prefs.boolForKey("useCustomSyncService")
+        if useCustomSyncService == true {
```
Nit: make this match the style of the others configurations please.

iron77 commented:
Hi guys,

First, great thanks for moving this issue forward!

Not sure if this is a good place but perhaps I'll try to ask: I now have installed Firefox 9.1 on my iOS 11 and I am wondering if custom FxA servers configuration have already been pushed into production, as I cannot find it anywhere in the settings? 

And if it is in production - is it possible that I do not see it because I never used Firefox Sync on this iOS device, so I don't have access "Account Settings" menu, instead having just a "Log in to Firefox" on the top of the Settings menu in iOS Firefox? Do I need to create a dummy Firefox sync account in official servers first to be passed in there and be able to change servers later?

And also, it seems like [our Bountysource pot](https://www.bountysource.com/issues/28349461-bounty-allow-usage-of-own-firefox-sync-token-server) have not been claimed yet. @vbudhram, do I understand correctly you're the one mainly responsible for making it happen? :)

chscherrer commented:
@jhreis pity your code didn't manage the latest release 10.0. Can't wait getting the possibility to define my own sync servers....

dprandzioch commented:
Just a question, is this already in the released version in the App Store?

yhager commented:
Can this be used with a private sync server, but still using accounts.firefox.com for the account management and auth? I have this setup for my desktop firefox, but not sure how to apply this to ios. I tried creating a `fxa-client-configuration` file with the same auth, oauth and profile servers as the default, and changed the tokenserver to mine. I serve that file from my own web server, but firefox still tries to `/signin` my web server (the one I specified in the advanced account screen).

dprandzioch commented:
Would also be really interested where to find it, but sadly there’s no reaction for over 4 months now...

iron77 commented:
@ctsrc @dprandzioch Yeah, I couldn't find it as well. It's little tricky, but fortunately it is deployed :)

> For Firefox for iOS version 9.0 or later:
        Go to Settings.
        Tap on the Version number 5 times.
        Tap on “Advance Account Settings”
        Enter your content-server URL
        Toggle “Use Custom Account Service” to on.

As per: https://mozilla-services.readthedocs.io/en/latest/howtos/run-fxa.html

Findus23 commented:
@iron77 
Am I correct that this only works when one is also using a custom account server?
So it isn't possible to use the firefox account service and just use a custom sync server (so just `identity.sync.tokenserver.uri` in the desktop and android version)

iron77 commented:
@Findus23 AFAIK, you're correct. I'm trying to work out an Ansible playbook that deploys all the microservices to a working FxA Account Server + Syncserver setup. If succeeded, I'm looking to share it at github so someone more knowledged can improve it further :). There are few already there but they don't work with latest FxA repos.

dprandzioch commented:
This is imo related to #3150 which is about using a self-hosted tokenserver without having to have a set up a full FxA stack that's only barely documented.

raphaelyancey commented:
Can't find the option in debug mode on Firefox 11.0 (9935) on iPad, is it gone ?

Otherwise, would love to see split fields to specify custom FxA and/or syncserver endpoints, as other said, because ones is most likely to run its own syncserver only instead of the whole FxA + syncserver stack.

rainforest1155 commented:
@raphaelyancey Try tapping 5 times on the version number in the settings. That should enable the debug mode allowing you to set up a custom FxA. There's just no option right now to just use a token syncserver - see #3150.

godfuture commented:
Will there be an option to simply use a custom token syncserver? 
The current menu for FxA custom server could be adjusted to split it in both auth and sync server URI. Then more use cases will be possible.

kriss6557 commented:
Hello,

I have set up a syncserver on my LAN and connected my laptop version of firefox to it (windows).I set value identity.sync.tokenserver.uri to http://syncserver.local:5000/token/1.0/sync/1.5I checked that synchonization fills up the database.

now  I want to do the same with my iphone, which is on the local WIFI and resolves correctly http://syncserver.local:5000/token/1.0/sync/1.5

I went in the debug mode to get the advanced menu  and set :

    custom FxA content server to https://accounts.firefox.com
    custom sync token server to http://syncserver.local:5000/token/1.0/sync/1.5

Then I connect to my account at https://accounts.firefox.com

I can see my iphone on the https://accounts.firefox.com page but synchronizing the phone doesn't work and always shows a past time.

I know my syncserver is not public and uses http but this works on my windows laptop (same subnet).

I'm not sure how to look at the logs on an iphone but if you find something wrong in my configuration please let me know.
Can you please help.
Thank you

johnwmail commented:
After enabled "debug mode" (tap five times on build number), 
I still can’t find this option. Is it available yet? Where do I find it?

SimonBasca commented:
@johnwmail you need to be signed out of FxA and the button will be below the "Sign in" button in Settings.

johnwmail commented:
> @johnwmail you need to be signed out of FxA and the button will be below the "Sign in" button in Settings.

Yes, I find the setting, after signed out, thanks.
